### ISSUE TITLE
[FEA] Stop working around Cython issues

### ISSUE THREAD

**Issue by vyasr** [rapidsai/cudf#14023]
**Is your feature request related to a problem? Please describe.**
cudf Cython code currently contains various workarounds for Cython limitations. With the Cython 3 release, many of these limitations have been lifted. Cython development is quite active right now, and now that we have migrated to Cython 3 we can also track new minor/patch version updates easily if new issues are fixed. This is a tracking issue of the various workarounds in cudf that we should look into removing as they are resolved upstream.

Due to the ongoing pylibcudf effort, in general it will not make sense to fix these issues in legacy cudf Cython code. We should instead endeavor to fix these issues in new pylibcudf code that will replace legacy Cython over time.

**Describe the solution you'd like**

- Enum class support
  - Added in Cython 3, but there are various workarounds currently in place. Once https://github.com/cython/cython/issues/5637 and https://github.com/cython/cython/issues/5609 are merged and released, we can update cudf to remove workarounds introduced in https://github.com/rapidsai/cudf/pull/13931.
- reference_wrapper
  - Currently we have [our own declarations](https://github.com/rapidsai/cudf/blob/branch-23.10/python/cudf/cudf/_lib/cpp/libcpp/functional.pxd). There's [an open PR](https://github.com/cython/cython/pull/5671) to add this to Cython, once it's merged we can remove our version
- Use of const in templates
  - There are issues in Cython parsing const inside templates. https://github.com/cython/cython/issues/3885 was resolved by https://github.com/cython/cython/pull/3886, but https://github.com/cython/cython/issues/4180 is still open. Based on my experimentation within cudf there are still some contexts where this is not resolved correctly. The current workaround inside cudf is to ctypedef a combined type (e.g. `ctypedef const scalar constscalar`). Once the bugs are fixed we should be able to get rid of the extraneous typedefs.
- optional
  - Support for std::optional was added in https://github.com/cython/cython/pull/3294 so we can use that anytime now
- make_unique
  - The fix we need for make_unique was completed in https://github.com/cython/cython/pull/5569, so we can use that as well

**Comment by vyasr** [comment]
reference_wrapper, optional, and make_unique are handled in #14382

**Comment by vyasr** [comment]
The outstanding work here is blocked on Cython improvements
- https://github.com/cython/cython/issues/4180 
- https://github.com/cython/cython/issues/5609

I'll close for now since I don't see a timeline for any of this work. There are inline comments indicating that we're working around Cython limitations for those, so they're nice-to-have fixes if and when Cython updates accordingly.
